rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,461 @@
+
+```rust
+#[cfg(kani)]
+fn mock_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(rand::random, mock_random)]
+fn random_cannot_be_zero() {
+    assert_ne!(rand::random::<u32>(), 0);
+}
+```
+
+Under this substitution, Kani has a single check, which proves that the assertion can fail. Verification time is 0.02 seconds.
+
+## User experience
+
+This feature is currently limited to stubbing functions and methods.
+We anticipate that the annotations we propose here could also be used for stubbing types, although the underlying technical approach might have to change.
+
+Stubs will be specified per harness; that is, different harnesses can use different stubs.
+This is one of the main design points.
+Users might want to mock the behavior of a function within one proof harness, and then mock it a different way for another harness, or even use the original function definition.
+It would be overly restrictive to impose the same stub definitions across all proof harnesses.
+A good example of this is compositional reasoning: in some harnesses, we want to prove properties of a particular function (and so want to use its actual implementation), and in other harnesses we want to assume that that function has those properties.
+
+Users will specify stubs by attaching the `#[kani::stub(<original>, <replacement>)]` attribute to each harness function.
+The arguments `original` and `replacement` give the names of functions/methods.
+They will be resolved using Rust's standard name resolution rules; this includes supporting imports like `use foo::bar as baz`.
+The attribute may be specified multiple times per harness, so that multiple (non-conflicting) stub pairings are supported.
+
+For example, this code specifies that the function `mock_random` should be used in place of the function `rand::random` and the function `my_mod::bar` should be used in place of the function `my_mod::foo` for the harness `my_mod::my_harness`:
+
+```rust
+#[cfg(kani)]
+fn mock_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+mod my_mod {
+
+    fn foo(x: u32) -> u32 { ... }
+
+    fn bar(x: u32) -> u32 { ... }
+
+    #[cfg(kani)]
+    #[kani::proof]
+    #[kani::stub(rand::random, super::mock_random)]
+    #[kani::stub(foo, bar)]
+    fn my_harness() { ... }
+
+}
+```
+
+We will support the stubbing of private functions and methods.
+While this provides flexibility that we believe will be necessary in practice, it can also lead to brittle proofs: private functions/methods can change or disappear in even minor version upgrades (thanks to refactoring), and so proofs that depend on them might have a high maintenance burden.
+In the documentation, we will discourage stubbing private functions/methods except if absolutely necessary.
+
+### Stub sets
+
+As a convenience, we will provide a macro `kani::stub_set` that allows users to specify sets of stubs that can be applied to multiple harnesses:
+
+```rust
+kani::stub_set! {
```

Comment:
nit/bike-shed: I think this syntax isn't ideal. How much control over it do you have? If I'm not mistaken you're suggesting that the first item in the list is the name of the stub set, right?

Can we do something like `kani::stub_set!(my_io_stubs) { ... the contents ... }` or similar?

Reply:
Exactly, the first item would be the name of the stub set. How about `kani::stub_set!(my_io_stubs, { ... the contents ... });`? We should be able to support that. I believe there needs to be a single outer set of parentheses or braces enclosing everything (the name, the contents, etc.).

Reply:
How about: `kani::stub_set! { my_io_stubs: ... }`? This would avoid having two types of delimiters (`{}` and `()`).

Reply:
Thanks @fzaiser! I think this is a pretty good solution. I'm going to go with it for now; we can always iterate on it.

Reply:
I am actually not a big fan of using `:` as a separator. It looks like a typo for me. I would prefer regular rust separators,  `,`, or an aggregator: `{}` or `()`. E.g.: like:

```rust
kani::stub_set!(my_io_stubs {
 ... the contents ... 
});
```

Reply:
Ideally we use something that rustfmt will accept too 🙂. It basically needs to look like an existing construct for it to work.

Reply:
@camshaft That's a good point about rustfmt. Apparently it [doesn't format code in macro invocations between braces](https://users.rust-lang.org/t/rustfmt-skips-macro-arguments/74807/17). This syntax (based on @celinval's) works well with rustfmt:

```
kani::stub_set!(my_io_stubs(
    stub(std::fs::read, my_read),
    stub(std::fs::write, my_write),
));
```